src/core/imap.ts

```diff
@@ -0,0 +1,130 @@
+import * as Imap from 'imap'
+import { ImapConfig } from '../model/imapConfig.model';
+import { Query } from '../model/Query.model';
+
+export class IMAP {
+
+    private imap = null;
+    private monitorId = null;
+    private query: Query = null;
+    private delayTime = 1000 * 5;
+
+    constructor(config: ImapConfig) {
+
+        this.imap = new Imap({
+            ...config,
+            authTimeout: 10000,
+            connTimeout: 30000,
+            tlsOptions: {
+                rejectUnauthorized: false
+            },
+
+        });
+    }
+
+    init() {
+        return new Promise((resolve, reject) => {
+
+            this.imap.connect();
+
+            this.imap.once("ready", function () {
+                console.log('%c---------------------  SUCCESSFUL IMAP CONNECTION  --------------------', 'color: Green');
+                resolve(this.imap);
+            });
+
+            this.imap.once('error', function (err: Error) {
+                reject(err);
+            });
+        });
+    }
+
+    setQuery(query: Query) {
+        this.query = query;
+    }
+
+    openBox(mailBox, readOnly = true) {
+        return new Promise((resolve, rejects) => {
+
+            this.imap.openBox(mailBox, readOnly, (err, box) => {
+                if (err) {
+                    rejects(err);
+                }
+                else {
+                    resolve(box);
+                }
+            });
+        });
+    }
+
+    search(criteria) {
+        return new Promise((resolve, rejects) => {
+            this.imap.search(criteria, (err, messages) => {
+                if (err) {
+                    rejects(err);
+                }
+                else {
+                    resolve(messages);
+                }
+            });
+        });
+    }
+
+    state() {
+
+        return new Promise((resolve, rejects) => {
+
+            if (!navigator.onLine) {
+                rejects('No internet Connection')
+            }
+            else if (!this.imap)
+                rejects('Please Re-login')
+            else if (this.imap.state == 'authenticated')
+                resolve('authenticated');
+            else {
+                this.imap.connect().then(() => {
+                    resolve('Reconnected successfully');
+                }, (err: Error) => {
+                    rejects(err);
+                });
+            }
+        })
+    }
+
+    monitor() {
+
+        let mailBox = null;
+        let criteria = null;
+
+        this.monitorId = setInterval(async () => {
+            //  
+            if (this.query) {
+
+                ({ mailBox, criteria } = this.query);
+
+                // Check if the connection is still stable.
+                await this.state().then(async () => {
```

Comment:
isn't this the same as
```
await this.state()
await this.openBox(mailBox)
let messages = await this.search(criteria)
...
```

Reply:
> isn't this the same as

Yes, it's the same thing.

> let messages = await this.search(criteria)

But if this method is used, a new variable will be created in each monitoring process.

However, there is another way to declare this variable above the monitoring and reassignment it every time.
Do I change it this way?

Reply:
> But if this method is used, a new variable will be created in each monitoring process.

Is it a problem?

Reply:
> Is it a problem?

I am not sure 100%, but I think it may cause memory leaks until the user logs out.
But I can declare the variable before monitoring to avoid this problem!

Reply:
A variable on its own does not use memory. It's the data that is returned from `this.search`. Declaring a variable in advance changes nothing.

Reply:
> A variable on its own does not use memory. It's the data that is returned from `this.search`. Declaring a variable in advance changes nothing.

OK, I get it now .
I have replaced 'then' with 'await'.